### PR TITLE
dotty: wrap to prevent installing all the files in profile

### LINKED ISSUE #31858: Dotty in conflict with Scala
## Issue description

I've [added](https://github.com/NixOS/nixpkgs/pull/31478) a new package `dotty` to nixpkgs, which represents the new version of Scala 3.0. 

It works great when installed, but unfortunately causes a conflict with previous versions of Scala.

Is it acceptable, given that we try to install different versions of Scala compiler? 

If not, I would be happy to fix this problem but I don't know how to proceed... Can anyone give me some hints?

## Steps to reproduce

```
nix-env -i scala
nix-env -i dotty
```

Output :
```
installing ‘dotty-0.4.0-RC1’
building path(s) ‘/nix/store/7cd0ff9hva3fs9rgvhsz85bghp98hvkh-user-environment’
Wide character in die at /nix/store/64jc9gd2rkbgdb4yjx3nrgc91bpjj5ky-buildenv.pl line 79.
collision between ‘/nix/store/vxzl4i68vgrfmdhrfcj61cs8b4iazfiq-dotty-0.4.0-RC1/lib/scala-xml_2.12-1.0.6.jar’ and ‘/nix/store/kyl3d46lgrqs9c2zk1q2j074d37l7rqv-scala-2.12.4/lib/scala-xml_2.12-1.0.6.jar’; use ‘nix-env --set-flag priority NUMBER PKGNAME’ to change the priority of one of the conflicting packages
builder for ‘/nix/store/kjbgvv335b7n393iykzmgaj557xjwana-user-environment.drv’ failed with exit code 2
error: build of ‘/nix/store/kjbgvv335b7n393iykzmgaj557xjwana-user-environment.drv’ failed
```

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.61, NixOS, 18.03.git.35e52ff (Impala)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 1.11.15`
 - channels(karol): `""`
 - channels(root): `""`
 - nixpkgs: `/etc/nixos/nixpkgs`

### PR DESCRIPTION
**Pull request by karolchmist**
###### Motivation for this change
Fixes  #31858 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by karolchmist** [comment]
@grahamcofborg build dotty

**Comment by karolchmist** [comment]
Pinging @yegortimoshenko, hello :) You've [helped me](31478) with the last PR of dotty suggesting wrapping the derivation to prevent the leaking of /bin, would you mind reviewing this PR? Thanks!

**Comment by yegortimoshenko** [comment]
@GrahamcOfBorg build dotty

**Comment by yegortimoshenko** [comment]
@grahamc: x86_64-darwin is having some issues!

File: pkgs/development/compilers/scala/dotty.nix [hunk 1]
```diff
@@ -1,46 +1,28 @@
...
 
-stdenv.mkDerivation rec {
-  version = "0.4.0-RC1";
-  name = "dotty-${version}";
-
-  src = fetchurl {
-    url = "https://github.com/lampepfl/dotty/releases/download/${version}/${name}.tar.gz";
-    sha256 = "1d1ab08b85bd6898ce6273fa50818de0d314fc6e5377fb6ee05494827043321b";
+let
+  self = {
```
Comment (yegortimoshenko) [inline pkgs/development/compilers/scala/dotty.nix]: This `self` only has one attribute and thus could be omitted:

```
let
    dotty-bare = callPackage ./dotty-bare.nix {
      inherit stdenv fetchurl makeWrapper jre;
    };
in
```

File: pkgs/development/compilers/scala/dotty-bare.nix [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  propagatedBuildInputs = [ jre ] ;
+  buildInputs = [ makeWrapper ] ;
+
+  installPhase = ''
+    mkdir -p $out
+    mv * $out
+  '';
+
+  fixupPhase = ''
+        for p in $out/bin/* ; do
```
Comment (yegortimoshenko) [inline pkgs/development/compilers/scala/dotty-bare.nix]: If `$out/bin/*` here were to be replaced with `$(find $out/bin -type f ! -name common)`, conditions below would not be required.

File: pkgs/development/compilers/scala/dotty.nix [hunk 3]
```diff
@@ -1,46 +1,28 @@
...
-       Dotty is a platform to try out new language concepts and compiler technologies for Scala.
-       The focus is mainly on simplification. We remove extraneous syntax (e.g. no XML literals),
-       and try to boil down Scala’s types into a smaller set of more fundamental constructs.
-       The theory behind these constructs is researched in DOT, a calculus for dependent object types.
-    '';
-    homepage = http://dotty.epfl.ch/;
-    license = licenses.bsd3;
-    platforms = platforms.all;
-    maintainers = [maintainers.karolchmist];
+  meta = with dotty-bare.meta; {
```
Comment (yegortimoshenko) [inline pkgs/development/compilers/scala/dotty.nix]:
```
  inherit (dotty-bare) meta;
```

**Review by lukateras — APPROVED** [review]
Other than minor syntactical nitpicks, this is great!

**Comment by karolchmist** [comment]
Thanks @yegortimoshenko , I've modified the code as you suggested.

**Pull request merged**